### PR TITLE
Functions ASB sample update

### PR DESCRIPTION
**Pull request by SeanFeldman**
Includes two samples now:
1. Statically created `FunctionEndpoint`.
2. `FunctionEndpoint` created using `IFunctionHostBuilder`.

Both samples demonstrate how to inject and use a custom component.

### CODE REVIEW

File: samples/previews/azure-functions/service-bus/sample.md [hunk 1]
```diff
@@ -45,6 +52,22 @@ To try the Azure Function:
 
 ## Code walk-through
 
+### `IFunctionHostBuilder` approach
+
+The NServiceBus endpoint configured using `IFunctionHostBuilder` is using the convention and is wired using `Startup` class like this:
+
+snippet: configuration-with-function-host-builder
+
+`FunctionEndpoin` is then injected into the function class:
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/sample.md:61]:
```suggestion
`FunctionEndpoint` is then injected into the function class:
```

**Review by timbussmann — COMMENTED** [review]
do we really want to use a single sample for this? There's a good chance that by default, all three projects will be configured as startup projects and unless the sender sends to different queues, it becomes a competing consumer problem between the two function hosts? Might be easier to go with two dedicated samples?

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/AzureFunctions.ASBTrigger.FunctionsHostBuilder.csproj [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    <LangVersion>7.3</LangVersion>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="..\AzureFunctions.ASBTrigger.Static\CustomComponent.cs" Link="CustomComponent.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="NServiceBus.AzureFunctions.ServiceBus" Version="0.3.0-*" />
+    <PackageReference Include="NServiceBus.Newtonsoft.Json" Version="2.*" />
+    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.*" />
+    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.*" />
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/AzureFunctions.ASBTrigger.FunctionsHostBuilder.csproj:14]: this shouldn't be necessary either, is it?

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/Startup.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            var customComponentInitializationValue = configurationRoot.GetValue<string>("CustomComponentValue");
+
+            return new CustomComponent(customComponentInitializationValue);
+        });
+
+        builder.UseNServiceBus(() =>
+        {
+            var configuration = new ServiceBusTriggeredEndpointConfiguration(AzureServiceBusTriggerFunction.EndpointName);
+
+            // optional: log startup diagnostics using Functions provided logger
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/Startup.cs:29]: not sure I understand this comment? The Logging will use the functions provided logger?

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/TriggerMessageHandler.cs [hunk 4]
```diff
@@ -6,11 +6,19 @@
...
+
+    public TriggerMessageHandler(CustomComponent customComponent)
+    {
+        this.customComponent = customComponent;
+    }
 
     public Task Handle(TriggerMessage message, IMessageHandlerContext context)
     {
         Log.Warn($"Handling {nameof(TriggerMessage)} in {nameof(TriggerMessageHandler)}");
+        Log.Warn($"Custom component returned: {customComponent.GetValue()}");
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/TriggerMessageHandler.cs:20]: should we change the warn to info?

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.Static/CustomComponent.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+
+public class CustomComponent
+{
+    readonly string initialValue;
+
+    public CustomComponent(string initialValue)
+    {
+        this.initialValue = $"{initialValue} {DateTime.UtcNow:O}";
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.Static/CustomComponent.cs:9]: what's the purpose of the timestamp?

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/Startup.cs (hunk 3 above)
Comment (SeanFeldman) [reply samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/Startup.cs:29]: We keep it as-is to show that logging diagnostics is possible and it's optional.

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/TriggerMessageHandler.cs (hunk 4 above)
Comment (SeanFeldman) [reply samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/TriggerMessageHandler.cs:20]: This is to make the logs from the handlers stand out and be more prominent among the other info logs added by function.

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.Static/CustomComponent.cs (hunk 5 above)
Comment (SeanFeldman) [reply samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.Static/CustomComponent.cs:9]: This is to demonstrate that the Singleton scope is actually working 🙂

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.Static/AzureServiceBusTriggerFunction.cs [hunk 6]
```diff
@@ -29,11 +30,16 @@ public class AzureServiceBusTriggerFunction
         // endpoint name, and connection strings are automatically derived from FunctionName and ServiceBusTrigger attributes
         var configuration = ServiceBusTriggeredEndpointConfiguration.FromAttributes();
 
-        configuration.UseSerialization<NewtonsoftSerializer>();
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.Static/AzureServiceBusTriggerFunction.cs:32]: given this has been removed, do we need the package reference in the csproj anymore?

File: samples/previews/azure-functions/service-bus/sample.md [hunk 7]
```diff
@@ -1,13 +1,17 @@
...
-reviewed: 2020-06-15
+reviewed: 2020-11-10
 component: ASBFunctions
 related:
  - previews/azure-functions-service-bus
 ---
 
 This sample shows how to host NServiceBus within an Azure Function, in this case, a function triggered by an incoming Service Bus message. This enables hosting message handlers in Azure Functions, gaining the abstraction of message handlers implemented using `IHandleMessages<T>` and also taking advantage of NServiceBus's extensible message processing pipeline.
 
+The sample demonstrates two approaches that achieve the same outcome:
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/sample.md:11]:
```suggestion
The sample demonstrates two configuration approaches that achieve the same outcome:
```

File: samples/previews/azure-functions/service-bus/sample.md [hunk 8]
```diff
@@ -1,13 +1,17 @@
...
+1. Using `IFunctionHostBuilder` and Microsoft DI container approach
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/sample.md:12]:
```suggestion
1. Integrating with the `IFunctionHostBuilder`, including the host managed DI container.
```

File: samples/previews/azure-functions/service-bus/sample.md [hunk 9]
```diff
@@ -1,13 +1,17 @@
...
+2. Static approach
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/sample.md:13]:
```suggestion
2. Configuring the endpoint inside the trigger class as static field.
```

File: samples/previews/azure-functions/service-bus/sample.md [hunk 10]
```diff
@@ -24,10 +28,16 @@ To create the queue with the Azure CLI, execute the following [Azure CLI](https:
...
-To use the sample, a valid Service Bus connection string must be configured in  2 locations:
+To use the sample, a valid Service Bus connection string must be provided in the `local.settings.json` file.
+
+## Sample structure
 
-* `AzureFunctions.Sender/local.settings.json`
-* `AzureFunctions.ASBTrigger/local.settings.json`
+The sample contains the following projects:
+- `AzureFunctions.Sender` - console application to generate a trigger message
+- `AzureFunctions.ASBTrigger.FunctionsHostBuilder` - implemenetation using `IFunctionHostBuilder` approach to host NServiceBus endpoint
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/sample.md:37]:
```suggestion
- `AzureFunctions.ASBTrigger.FunctionsHostBuilder` - Using `IFunctionHostBuilder` approach to host the NServiceBus endpoint
```

File: samples/previews/azure-functions/service-bus/sample.md [hunk 11]
```diff
@@ -24,10 +28,16 @@ To create the queue with the Azure CLI, execute the following [Azure CLI](https:
...
+- `AzureFunctions.ASBTrigger.Static` - implementation using static approach to host NServiceBus endpoint
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/sample.md:38]:
```suggestion
- `AzureFunctions.ASBTrigger.Static` - Using static approach to host the NServiceBus endpoint
```

File: samples/previews/azure-functions/service-bus/sample.md [hunk 12]
```diff
@@ -55,7 +81,10 @@ snippet: Function
 
 Meanwhile, the message handlers for `TriggerMessage` and `FollowUpMessage`, also hosted within the Azure Functions project, are normal NServiceBus message handlers, which are also capable of sending messages themselves.
 
+### Handlers
+
+Both approaches use the same handlers code, with a `CustomDependency` passed in.
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/sample.md:86]:
```suggestion
Both approaches use the same message handlers, with a `CustomDependency` passed in.
```

**Review by timbussmann — COMMENTED** [review]
some small suggestions.

**Comment by SeanFeldman** [comment]
All suggestions are addressed.

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/AzureFunctions.ASBTrigger.FunctionsHostBuilder.csproj [hunk 13]
```diff
@@ -0,0 +1,27 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
+    <LangVersion>7.3</LangVersion>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="..\AzureFunctions.ASBTrigger.Static\CustomComponent.cs" Link="CustomComponent.cs" />
```
Comment (timbussmann) [inline samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/AzureFunctions.ASBTrigger.FunctionsHostBuilder.csproj:8]: we can also include the message handlers to remove duplicates. See https://github.com/Particular/docs.particular.net/pull/5115

**Comment by SeanFeldman** [comment]
`prerelease.txt` removed on purpose.

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.Static/AzureFunctions.ASBTrigger.Static.csproj [hunk 14]
```diff
@@ -5,7 +5,7 @@
     <LangVersion>7.3</LangVersion>
   </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="NServiceBus.AzureFunctions.ServiceBus" Version="0.2.*" />
+    <PackageReference Include="NServiceBus.AzureFunctions.ServiceBus" Version="0.3.0-*" />
```
Comment (SeanFeldman) [inline samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.Static/AzureFunctions.ASBTrigger.Static.csproj:8]:
```suggestion
    <PackageReference Include="NServiceBus.AzureFunctions.ServiceBus" Version="0.3.*" />
```

File: samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/AzureFunctions.ASBTrigger.FunctionsHostBuilder.csproj [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="NServiceBus.AzureFunctions.ServiceBus" Version="0.3.0-*" />
```
Comment (SeanFeldman) [inline samples/previews/azure-functions/service-bus/ASBFunctions_0/AzureFunctions.ASBTrigger.FunctionsHostBuilder/AzureFunctions.ASBTrigger.FunctionsHostBuilder.csproj:11]:
```suggestion
    <PackageReference Include="NServiceBus.AzureFunctions.ServiceBus" Version="0.3.*" />
```

**Review by SeanFeldman — COMMENTED** [review]
Adding two suggestions to be merged when we're ready to release.

**Pull request merged**